### ISSUE TITLE
Severe error in filtering procedure - critical memory corruption

### ISSUE THREAD

**Issue by vmarkovtsev** [vinniefalco/DSPFilters#15]
Cascade.h, near line 181:

```c++
  // Process a block of samples in the given form
  template <class StateType, typename Sample>
  void process (int numSamples, Sample* dest, StateType& state) const
  {
    while (--numSamples >= 0)
      *dest++ = state.process (*dest, *this);
  }
```

There is a very stupid error: the order of calculation of *dest++ and *dest is unspecified (see [this question on StackOverflow](http://stackoverflow.com/questions/1831316/is-this-ptr-ptr-a-undefined-behavior)). It seems that with Visual C++ it does the right thing, but g++ under Linux compiles the other way, resulting in the first sample to be skipped. Moreover, this actually leads to memory corruption in the end. I recommend the following fix:

```c++
  // Process a block of samples in the given form
  template <class StateType, typename Sample>
  void process (int numSamples, Sample* dest, StateType& state) const
  {
    while (--numSamples >= 0) {
      *dest = state.process (*dest, *this);
      dest++;
    }
  }
```

**Comment by galchinsky** [comment]
Fixed in #10

**Comment by loseyourself-gan** [comment]
@vmarkovtsev hello, can you give me an example how to use the process function, my code is
```
    const int order = 2;
    Dsp::Butterworth::LowPass<order> lowPassFilter;

    double sampleRate = 44100.0;   
    double cutoffFrequency = 1000.0; 

    lowPassFilter.setup(order, sampleRate, cutoffFrequency);

    lowPassFilter.process()
```
thank you very much
